### PR TITLE
Extend backend "ClientTypes"

### PR DESCRIPTION
**Pull request by StachuDotNet**
Changelog:

```
Internal improvements:
- Extracted most client-facing types to central "ClientTypes" project to allow for safe adjustments of those types without leaking into our domain logic
```

---

Our backend passes data back and forth with our client in a few different ways:
- via API request/response payloads
- via WASM-compiled code used to support "Analysis"
- via `ApiServer/UI.fs` which injects JSON into `ui.html` dynamically
- via Pusher.com 'pushes'

Presently, all of this communication is done via JSON payloads, serialized largely by our `Prelude.Vanilla` System.Text.Json serializer. The types in `fsharp-backend` corresponding to those JSON payloads are spread throughout our codebase, resulting in us serializing "internal" types.

That said, there's been some effort in the past to create "client types" in a `ClientTypes` project - the idea is to 'map' between internal types and external types, such that we may safely update either without breaking the communication between the Client and the Server. Generally, I've seen this idea referred to as "DTOs - Data Transfer Objects." Our ClientType definitions/setup is currently quite thin, and only covers Analysis and several APIs.

As is, it's unsafe to make many sorts of adjustments to our internal F# types (e.g. removing or renaming fields), as the serialized payload wouldn't be parse-able in our ReScript code, or vice versa.

The intent of this PR is to consolidate most (if not all) of the types used in client-server communication, to reduce/remove the current risks involved in updating internal types.

---

The specific cause of this work being done is an ongoing effort to "rebrand" our "Patterns" as "Match Patterns," as we'd like to shortly implement "let patterns" and so our current 'branding' would be confusing. Specifically, these existing "patterns" are serialized in a style of "`PBlank`," (corresponding to a 'blank' match pattern) and we'd like them to instead be serialized as "`MPBlank`". Our `client` is set up to handle this change, (our Match Pattern deserializer is able to parse both `MPBlank` and `PBlank`) but in order to finish this work, we need to be able to safely pass the new "`MPBlank`"-style payloads from `client` to our backend. Our backend's "AddOpV1" API endpoint request payload includes Match Patterns, and the types referenced are "internal" (not client types).

So, the motivation behind this PR is to ensure all types referenced by the "AddOpV1" are ClientTypes, where we may safely have both "`MPBlank`" _and_ "`PBlank`"-style definitions for these patterns, and translate those appropriately to a single "`MPBlank`" case in our "internal" types. (so the rename doesn't bleed into our domain logic). We could have gotten around this by writing a specialized serializer for the MatchPattern type, but moving everything to ClientTypes seems more beneficial long-term.

---

This work involved:
- breaking down the ClientTypes project into multiple modules (RuntimeTypes and AnalysisTypes)
- cloning many internal types (anything referenced by the noted means of client-server communication) into ClientTypes
  ProgramTypes, Ops, Worker, Secrets, etc,
- migrating all existing serializable types to ClientTypes
  - API requests/responses, Pusher.com payloads, ui.html-injected data
- creating additional projects to convert/map between ClientTypes and internal types
  - `ClientTypes2ExecutionTypes` handles conversions for our _core_ types defined in `LibExecution`
  - `ClientTypes2BackendTypes` handles types elsewhere, namely in `LibBackend` which is responsible for the surrounding infrastructure
- along the way, adjust/correct our testing around stable serializations of these types
  - `Serialization.Tests.fs`: reference the new types in tests
  - `TestJsonEncoding.res`: mostly, file names have simply adjusted due to updates in `Serialization.Tests.fs`

---

At this point, the PR is aimed to be a no-op; no functional changes. As far as I can tell, that is the case.

I have some outstanding questions/concerns/todos:
- given these refactors, it's not clear to me where `Json.Vanilla.allow<CTApi.Ops.AddOpV1.Request> "ApiServer.AddOps"` calls should be made, to explicitly allow serializations of various types. Previously these types were generally 'registered' in the assembly which needed them, but I've been starting to think that this might make more sense in an `init` within `ClientTypes` itself
- I'm not sure if I need to add a `ClientTypes.Init.init` call to `DataTests`, `ExecHost`, and `CronChecker` projects.
- `Serialization.Tests.fs` is feeling a bit unwieldy - maybe we should break this down into 2 files, one for client-server serializations, and one for DB-bound serializations
  it feels a bit awkward/confusing to me right now that many of the static "Values" in that file are "client types" and many are "internal" types that we map to "serialized types". Maybe it was a mistake for me to update those static Values to ClientTypes?

---

This is a bit non-exhaustive - there are some tests around our Vanilla serialize still in `Serialization.Tests.fs` that reference types _outside_ of `ClientTypes`. Likely in another PR, I need to figure out where these are used and what to do about them.

### CODE REVIEW

File: client/src/analysis/AnalysisTypes.res [hunk 1]
```diff
@@ -387,6 +387,8 @@ module FourOhFour = {
   }
 }
 
+// TODO: these are used outside of analysis (e.g. in APIWorkers) so the type def
+// should probably move?
 module WorkerState = {
```
Comment (StachuDotNet) [inline client/src/analysis/AnalysisTypes.res:392]: Happy to leave this for another PR, but `WorkerState` types are used both in "analysis" and during initial load, so I think we should move these defs somewhere

File: fsharp-backend/src/ClientTypes/ClientProgramTypes.fs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    | EveryDay
+    | EveryWeek
+    | EveryFortnight
+    | EveryHour
+    | Every12Hours
+    | EveryMinute
+
+  // TODO This comment seems...suspicious. I wonder if this can be removed or something? @pbiggar
+  // We need to keep the IDs around until we get rid of them on the client
+  type ids = { moduleID : id; nameID : id; modifierID : id }
```
Comment (StachuDotNet) [inline fsharp-backend/src/ClientTypes/ClientProgramTypes.fs:121]: Happy to leave this for a follow-up PR, but "We need to keep the IDs around until we get rid of them on the client" felt like we might be able to remove this from ClientTypes - for now, I'd just copied it straight from ProgramTypes.fs

File: fsharp-backend/src/ClientTypes/ClientRuntimeTypes.fs [hunk 3]
```diff
@@ -0,0 +1,141 @@
+/// Runtime Types used for client-server communication so we may update backend
+/// types without affecting APIs.
+///
+/// These should all directly match `RuntimeTypes.res` in `client`.
+/// See `RuntimeTypes.fs` for documentation of these types.
```
Comment (StachuDotNet) [inline fsharp-backend/src/ClientTypes/ClientRuntimeTypes.fs:5]: This file is just a copy/paste from the previous `ClientTypes.fs`

File: fsharp-backend/src/ClientTypes/README.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# ClientTypes
+
+Any times that we use to communicate with our 'client'/editor:
+- via APIs (request/response payloads)
+- via Wasm (in Analysis)
+- via Pusher.com pushes
```
Comment (StachuDotNet) [inline fsharp-backend/src/ClientTypes/README.md:6]: TODO: I need to expand upon this README a bit

File: fsharp-backend/src/ClientTypes2BackendTypes/ClientTypes2BackendTypes.fsproj [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    <ProjectReference Include="../LibBackend/LibBackend.fsproj" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="paket.references" />
+    <Compile Include="Ops.fs" />
+    <Compile Include="StaticDeploy.fs" />
+    <Compile Include="Authorization.fs" />
+    <Compile Include="Worker.fs" />
+    <Compile Include="Pusher.fs" />
+    <Compile Include="UI.fs" />
```
Comment (StachuDotNet) [inline fsharp-backend/src/ClientTypes2BackendTypes/ClientTypes2BackendTypes.fsproj:26]: Alternatively, I could see us naming these files `ClientTypes2UI.fs` - wasn't sure what's best.

File: fsharp-backend/src/CronChecker/CronChecker.fs [hunk 6]
```diff
@@ -36,6 +36,7 @@ let main _ : int =
     LibService.Init.init name
     LibExecution.Init.init ()
     Telemetry.Console.loadTelemetry name Telemetry.DontTraceDBQueries
+    //? ClientTypes.Init.init name
```
Comment (StachuDotNet) [inline fsharp-backend/src/CronChecker/CronChecker.fs:39]: TODO: figure out if this should be removed/uncommented.

File: fsharp-backend/src/ExecHost/ExecHost.fs [hunk 7]
```diff
@@ -194,6 +194,7 @@ let main (args : string []) : int =
       (LibBackend.Init.init LibBackend.Init.WaitForDB name).Result
     let result = (run options).Result
     LibService.Init.shutdown name
+    //? ClientTypes.Init.init name
```
Comment (StachuDotNet) [inline fsharp-backend/src/ExecHost/ExecHost.fs:197]: TODO: figure out if this should be removed/uncommented.

File: fsharp-backend/tests/DataTests/DataTests.fs [hunk 8]
```diff
@@ -256,6 +256,7 @@ let main args =
   LibService.Telemetry.Console.loadTelemetry
     "DataTests"
     LibService.Telemetry.DontTraceDBQueries
+  //? ClientTypes.Init.init name
```
Comment (StachuDotNet) [inline fsharp-backend/tests/DataTests/DataTests.fs:259]: TODO: figure out if this should be removed/uncommented.

**Comment by StachuDotNet** [comment]
This is ready for an initial review. I tried my best to break the commits down well and add appropriate commentary.
That said, it's a rather large PR, so happy to 1:1 on this.

File: backend/serialization/vanilla_LibBackend-Pusher_LibBackend-Pusher-AddOpEventTooBigPayload_simple.json [hunk 9]
```diff
@@ -1,7 +0,0 @@
-{
```
Comment (pbiggar) [inline backend/serialization/vanilla_LibBackend-Pusher_LibBackend-Pusher-AddOpEventTooBigPayload_simple.json:1]: Unclear why this is gone.

File: fsharp-backend/src/ClientTypes/ClientTypes.fsproj [hunk 10]
```diff
@@ -11,12 +11,24 @@
     <RuntimeIdentifier>linux-x64</RuntimeIdentifier>
   </PropertyGroup>
   <ItemGroup>
-    <None Include="paket.references" />
-    <Compile Include="ClientTypes.fs" />
+    <ProjectReference Include="../Prelude/Prelude.fsproj" />
```
Comment (pbiggar) [inline fsharp-backend/src/ClientTypes/ClientTypes.fsproj:14]: This is a really great static guarantee!

File: fsharp-backend/src/ApiServer/ApiServer.fs [hunk 11]
```diff
@@ -222,40 +223,42 @@ let run (packages : Packages) : unit =
...
-  Json.Vanilla.allow<Traces.TraceDataV1.Params> "ApiServer.Traces"
-  Json.Vanilla.allow<Traces.TraceDataV1.T> "ApiServer.Traces"
-  Json.Vanilla.allow<Tunnels.Register.Params> "ApiServer.Tunnels"
-  Json.Vanilla.allow<Tunnels.Register.T> "ApiServer.Tunnels"
-  Json.Vanilla.allow<Workers.Scheduler.Params> "ApiServer.Workers"
-  Json.Vanilla.allow<Workers.Scheduler.T> "ApiServer.Workers"
-  Json.Vanilla.allow<Workers.WorkerStats.Params> "ApiServer.Workers"
-  Json.Vanilla.allow<Workers.WorkerStats.T> "ApiServer.Workers"
+  // allow serialization of all API request/response payload types
+  Json.Vanilla.allow<CTApi.Ops.AddOpV1.Request> "ApiServer.AddOps"
```
Comment (pbiggar) [inline fsharp-backend/src/ApiServer/ApiServer.fs:227]: Yeah, I guess these make more sense in the ClientTypes initializer.

File: fsharp-backend/src/ClientTypes/ClientProgramTypes.fs (hunk 2 above)
Comment (pbiggar) [reply fsharp-backend/src/ClientTypes/ClientProgramTypes.fs:121]: This comment is still valid I'm pretty sure.

File: fsharp-backend/tests/Tests/Serialization.Tests.fs [hunk 12]
```diff
@@ -606,15 +624,15 @@ module GenericSerializersTests =
...
 
       v<LibExecution.DvalReprInternalNew.RoundtrippableSerializationFormatV0.Dval>
         "complete"
         (testDval
          |> LibExecution.DvalReprInternalNew.RoundtrippableSerializationFormatV0.fromRT)
 
-      v<LibExecution.ProgramTypes.Oplist> "complete" testOplist
-      v<LibExecution.ProgramTypes.Handler.T> "simple" testHttpHandler
+      v<LibExecution.ProgramTypes.Oplist> "complete" ProgramTypes.testOplist
+      v<LibExecution.ProgramTypes.Handler.T> "simple" ProgramTypes.testHttpHandler
```
Comment (pbiggar) [inline fsharp-backend/tests/Tests/Serialization.Tests.fs:635]: Surely this isn't allowed serialize anymore?

File: fsharp-backend/tests/Tests/Serialization.Tests.fs [hunk 13]
```diff
@@ -892,16 +922,17 @@ module GenericSerializersTests =
...
                         nameID = 8UL
-                        typ = Some(PT.TInt)
+                        typ = Some(CT2Program.DType.toCT PT.TInt)
                         typeID = 9UL } ]
                   version = 1 } ]
-            userFns = testUserFunctions
-            userTypes = testUserTypes
-            packageFns = [ testPackageFn ]
+            userFns =
+              List.map CT2Program.UserFunction.toCT ProgramTypes.testUserFunctions
```
Comment (pbiggar) [inline fsharp-backend/tests/Tests/Serialization.Tests.fs:933]: It seems the tests here are sometimes for ClientTypes and sometimes for converted types. It feels like there should be tests for roundtrips from real types, right? Maybe also roundtrips from ClientTypes?

**Review by pbiggar — CHANGES_REQUESTED** [review]
This looks great.

My one suggestion would be to make the testing more thorough - we currently don't have full end-to-end analysis including the conversion functions, and there's so much conversion that typos are easy to sneak in (both now and in the future).

File: fsharp-backend/src/ClientTypes/ClientProgramTypes.fs (hunk 2 above)
Comment (StachuDotNet) [reply fsharp-backend/src/ClientTypes/ClientProgramTypes.fs:121]: The comment is valid. I was wondering if `ids` could be removed, as "until we get rid of them on the client" implies that they're not _useful_ in the client.
Do we get rid of them in the client? If so, where? It seems we might not want to get rid of them in the client in any case - don't we need those IDs (at least one of them?) to reference them for updates?

File: backend/serialization/vanilla_LibBackend-Pusher_LibBackend-Pusher-AddOpEventTooBigPayload_simple.json (hunk 9 above)
Comment (StachuDotNet) [reply backend/serialization/vanilla_LibBackend-Pusher_LibBackend-Pusher-AddOpEventTooBigPayload_simple.json:1]: We don't yet handle the case of the Pusher.com AddOpV1 payload being too big to push to the client.

```
let tooBigPayload = { tlids = tlids } |> Json.Vanilla.serialize
// CLEANUP: when changes are too big, notify the client to reload them. We'll
// have to add support to the client before enabling this. The client would
// reload after this.
// push canvasID "addOpTooBig" tooBigPayload
```

is the code before changes, and we have the equivalent (just swallow the issue) in the new code.

Since we don't yet serialize it, I figured we didn't need to register the serializer, nor test for how it serializes.

File: fsharp-backend/tests/Tests/Serialization.Tests.fs (hunk 12 above)
Comment (StachuDotNet) [reply fsharp-backend/tests/Tests/Serialization.Tests.fs:635]: It turns out they are.
- `LibExecution.ProgramTypes.Handler.T` seems to be used in a canvas-clone test, and the .allow is registered in `Tests.fs` just for that
- `LibExecution.ProgramTypes.Oplist` is used in `Canvas`'s `loadJsonFromDisk`


As I look through other non-ClientTypes types, I seem to find similar edge cases.

**Comment by StachuDotNet** [comment]
> My one suggestion would be to make the testing more thorough

Could we leave this for a follow-up PR? I'm bored/tired of ClientTypes at this point, and the PR as-is feels like a net positive.

File: fsharp-backend/tests/Tests/Serialization.Tests.fs (hunk 13 above)
Comment (StachuDotNet) [reply fsharp-backend/tests/Tests/Serialization.Tests.fs]: done!

File: fsharp-backend/src/ClientTypes/ClientProgramTypes.fs (hunk 2 above)
Comment (pbiggar) [reply fsharp-backend/src/ClientTypes/ClientProgramTypes.fs]: We need them while we update them via the BlankOr mechanism. When we extend fluid to handler specs, I expect we won't use IDs in handler specs anymore.

File: fsharp-backend/src/LibAnalysis/LibAnalysis.fs [hunk 14]
```diff
@@ -94,23 +93,26 @@ module Eval =
...
-let performAnalysis (args : CTA.PerformAnalysisParams) : Task<CTA.AnalysisEnvelope> =
+let performAnalysis
+  (args : CTAnalysis.PerformAnalysisParams)
+  : Task<CTAnalysis.AnalysisEnvelope> =
   let runAnalysis
     (tlid : tlid)
-    (traceID : CTA.TraceID)
-    (traceData : CTA.TraceData.T)
+    (traceID : CTAnalysis.TraceID)
+    (traceData : CTAnalysis.TraceData)
```
Comment (pbiggar) [inline fsharp-backend/src/LibAnalysis/LibAnalysis.fs:106]: It seems odd that these take clients types for some arguments. I think if we're going to convert from clients types in one place, we should do the conversion for all the arguments in that place.

File: fsharp-backend/src/ClientTypes/Converters.fs [hunk 15]
```diff
@@ -0,0 +1,19 @@
+module ClientTypes.Converters
+
+module STJ =
```
Comment (pbiggar) [inline fsharp-backend/src/ClientTypes/Converters.fs:3]: Perhaps there's an opportunity to get rid of this. We could argue that the ClientType here is really a `string` and the conversion functions should do the parsing, rather than rely on this (I feel this particular thing is annoying to learn and keep track of)

**Review by pbiggar — CHANGES_REQUESTED** [review]
This looks fantastic, nice job!

A couple of small suggestions and we should be good to merge.

File: client/src/analysis/AnalysisTypes.res [hunk 16]
```diff
@@ -387,6 +387,8 @@ module FourOhFour = {
   }
 }
 
+// TODO: these are used outside of analysis (e.g. in APIWorkers) so the type def
```
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply client/src/analysis/AnalysisTypes.res:390]: The `WorkerState` module/type isn't just used in Analysis (as implied by its location in `AnalysisTypes.res`).
It's also used in the "initial load" API payload, and is referenced by `APIWorkers` file.

Just thinking to move the module definition somewhere else, but I'm not sure where that would be.

File: fsharp-backend/src/LibBackend/Pusher.fs [hunk 17]
```diff
@@ -35,78 +45,46 @@ type EventTooBigEvent = { eventName : string }
...
-  push canvasID "new_404" (Json.Vanilla.serialize f404)
+        let! (_ : PusherServer.ITriggerResult) =
+          client.TriggerAsync(channel, event.EventName, event.Payload)
 
+        return ()
+      })
 
-let pushNewStaticDeploy (canvasID : CanvasID) (asset : StaticAssets.StaticDeploy) =
-  push canvasID "new_static_deploy" (Json.Vanilla.serialize asset)
+  let serialized = eventSerializer event
```
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply fsharp-backend/src/LibBackend/Pusher.fs:72]: good idea!

File: fsharp-backend/src/ClientTypes/Converters.fs (hunk 15 above)
Comment (StachuDotNet) [reply fsharp-backend/src/ClientTypes/Converters.fs:3]: I don't understand what you mean here. Could you please rephrase or briefly chat?

File: fsharp-backend/src/Analysis/Analysis.fs [hunk 18]
```diff
@@ -1,26 +1,21 @@
...
-module DvalReprInternalDeprecated = LibExecution.DvalReprInternalDeprecated
-
-module CTA = ClientTypes.Analysis
-
-open System
-open System.Reflection
-
 #nowarn "988"
 
+module Init =
```
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply fsharp-backend/src/Analysis/Analysis.fs:13]: I forget why the init is in LibAnalysis actually. I'll try throwing it into Analysis and see what breaks, if anything.

[the remark this answers is not in the document]
Comment (StachuDotNet) [reply fsharp-backend/src/Analysis/Analysis.fs:13]: Ah. Tests.fs calls on `LibAnalysis.initSerializers ()`. I think if I try moving those to Analysis, I get some .net WASM compile issue.

File: fsharp-backend/src/LibBackend/Pusher.fs (hunk 17 above)
[the remark this answers is not in the document]
Comment (StachuDotNet) [reply fsharp-backend/src/LibBackend/Pusher.fs]: This is done

File: fsharp-backend/src/LibAnalysis/LibAnalysis.fs (hunk 14 above)
Comment (StachuDotNet) [reply fsharp-backend/src/LibAnalysis/LibAnalysis.fs]: I agree. I ended up extracting this mapping out to `CT2Analysis`, including the bits where we map from PTs to RTs. I think this turned out much nicer.

File: fsharp-backend/src/ClientTypes/Converters.fs (hunk 15 above)
Comment (StachuDotNet) [reply fsharp-backend/src/ClientTypes/Converters.fs:3]: Never mind - I understand now.

File: fsharp-backend/src/ExecHost/ExecHost.fs [hunk 19]
```diff
@@ -207,16 +207,6 @@ let initSerializers () =
   Json.Vanilla.allow<CTPusher.Payload.New404> "Pusher"
   Json.Vanilla.allow<CTPusher.Payload.AddOpV1> "Pusher"
   //Json.Vanilla.allow<CTPusher.Payload.AddOpV1PayloadTooBig> "Pusher" // this is so-far unused
-  Json.Vanilla.registerConverter (ClientTypes.Converters.STJ.WorkerStateConverter())
-  Json.Vanilla.allow<CTPusher.Payload.UpdateWorkerStates> "Pusher"
-
-  // allow serialization of types used in Pusher.com payloads
```
Comment (StachuDotNet) [inline fsharp-backend/src/ExecHost/ExecHost.fs]: (apparently I had created duplicate `.allow` statements here for pusher.com payloads)

**Comment by StachuDotNet** [comment]
This has been rebased to latest - ready for another review.
